packages/falso/src/lib/core/core.ts

```diff
@@ -11,21 +12,94 @@ type Return<T, O extends FakeOptions> = [O] extends [never]
   : T;
 
 export function fake<T, Options extends FakeOptions>(
-  data: T[] | ((i: number) => T),
-  options?: Options
+  data: T[] | (() => T),
+  options?: Options,
+  comparisonFunction: (item: T, items: T[]) => boolean = checkUniquePrimitive
 ): Return<T, Options> {
-  const dataSource = Array.isArray(data) ? () => randElement(data) : data;
+  if (Array.isArray(data)) {
+    return fakeFromArray(data, options) as any;
+  }
+
+  return fakeFromFunction(data, comparisonFunction, options) as any;
+}
+
+export function fakeFromFunction<T, Options extends FakeOptions>(
+  data: () => T,
+  comparisonFunction: (item: T, items: T[]) => boolean,
+  options?: Options
+) {
+  if (!options?.length) {
+    return data();
+  }
+
+  const priority = options?.priority ?? 'length';
+
+  if (priority === 'length') {
+    return Array.from({ length: options.length }, (_, index) => data());
+  }
+
+  const items: T[] = [];
+
+  let attempts = 0;
+  const maxAttempts = options.length * 2;
+
+  while (items.length < options.length && attempts < maxAttempts) {
+    const item = data();
+
+    if (!comparisonFunction(item, items)) {
+      items.push(item);
+    }
 
+    attempts++;
+  }
+
+  return items;
+}
+
+export function fakeFromArray<T, Options extends FakeOptions>(
+  data: T[],
+  options?: Options
+) {
   if (!options?.length) {
-    return dataSource(0) as any;
+    return randElement(data);
+  }
+
+  const priority = options?.priority ?? 'length';
+
+  if (priority === 'length') {
+    return Array.from({ length: options.length }, () => randElement(data));
   }
 
-  return Array.from({ length: options.length }, (_, index) =>
-    dataSource(index)
-  ) as any;
+  const clonedData: T[] = JSON.parse(JSON.stringify(data));
+  const newArray: T[] = [];
+
+  while (clonedData.length && newArray.length !== options.length) {
+    const randomIndex = Math.floor(random() * clonedData.length);
+    const item = clonedData[randomIndex];
+
+    newArray.push(item);
+    clonedData.splice(randomIndex, 1);
+  }
+
+  return newArray;
 }
 
-export function randElement<T>(arr: T[]) {
+export const checkUniquePrimitive: <T>(item: T, items: T[]) => boolean = (
+  item,
+  items
+) => items.includes(item);
+
+export const checkUniqueDate: (date: Date, dates: Date[]) => boolean = (
+  date,
+  dates
+) => dates.some((d) => d.valueOf() === date.valueOf());
+
+export const checkUniqueObjectWithId: <T extends { id: string }>(
+  item: T,
+  items: T[]
+) => boolean = (item, items) => items.some((i) => i.id === item.id);
```

Comment:
I think you can create a more generic verifier which accepts a list of keys.
This will also be useful in the flight, credit card, address etc